algorithms/src/polycommit/sonic_pc/data_structures.rs

```diff
@@ -476,16 +514,54 @@ impl<E: PairingEngine> VerifierKey<E> {
             .as_ref()
             .and_then(|v| v.binary_search_by(|(d, _)| d.cmp(&degree_bound)).ok().map(|i| v[i].1.clone()))
     }
-}
 
-impl<E: PairingEngine> VerifierKey<E> {
     pub fn max_degree(&self) -> usize {
         self.max_degree
     }
 
     pub fn supported_degree(&self) -> usize {
         self.supported_degree
     }
+
+    pub fn union<'a, T: IntoIterator<Item = &'a Self> + Clone>(verifier_keys: T) -> Self {
+        let mut bounds_seen = HashSet::<usize>::new();
+        let mut bounds_and_neg_powers = vec![];
+        let mut bounds_and_prepared_neg_powers = vec![];
+        let mut biggest_vk: Option<&VerifierKey<E>> = None;
+        for vk in verifier_keys {
+            if biggest_vk.is_none() || biggest_vk.unwrap().supported_degree() < vk.supported_degree() {
+                biggest_vk = Some(vk);
+            }
+            let new_bounds = vk.degree_bounds_and_neg_powers_of_h.clone().unwrap();
+            let new_prep_bounds = vk.degree_bounds_and_prepared_neg_powers_of_h.clone().unwrap();
+            assert!(new_bounds.len() == new_prep_bounds.len());
+            for ((bound, neg_powers), (_, prep_neg_powers)) in new_bounds.into_iter().zip(new_prep_bounds) {
+                if bounds_seen.insert(bound) {
+                    bounds_and_neg_powers.push((bound, neg_powers));
+                    bounds_and_prepared_neg_powers.push((bound, prep_neg_powers));
+                }
```

Comment:
I think we can replace this with a use of the `entry` API again

Reply:
Hmm I'm skeptical. For context: `bounds_seen` is a `HashSet` while `bounds_and_neg_powers` and `bounds_and_prepared_neg_powers` are currently vectors. 

Whereas in `ahp.rs` we use the `map.entry()` to *always* return a value, sometimes cached - in this case we want to only collect and push values if the key is new. Using two `map.entry()`s could look cleaner here, but (1) we wouldn't be using the `entry` API properly as we don't care about retrieving the value and (2) a couple lines down it will be more involved to sort and dedup the map's values.

Reply:
Ah, I see. Sounds good!